lib/HLSL/LiveValueAnalysis.cpp

```diff
@@ -0,0 +1,1323 @@
+bool LiveValueAnalysis::canRemat(Instruction* inst) {
+  if( isa<GetElementPtrInst>( inst ) )
+    return true;
+  else if( isa<AllocaInst>( inst ) )
+    return true;
+  else if( isa<LoadInst>( inst ) )
+    return true;
+
+  if( CallInst* call = dyn_cast<CallInst>(inst) ) {
+    if( call->getCalledFunction()->getName().startswith( "dx.op.bufferLoad" ) && !usesUavResource( call ) )
+      return true;
+    else if (call->getCalledFunction()->getName().startswith("dx.op.textureLoad") && !usesUavResource(call))
+      return true;
+    else if (call->getCalledFunction()->getName().startswith("dx.op.rawBufferLoad") && !usesUavResource(call))
+      return true;
+    else if (call->getCalledFunction()->getName().startswith("dx.op.createHandleForLib"))
+      return true;
+  }
+
+  return false;
+}
+
+// Full path locations can be complicated and some projects have strange behavior. Customize accordingly.
+std::string LiveValueAnalysis::formatSourceLocation( const std::string& m_rootPath, const std::string& Location, int lineNumber )
+{
+  std::string fullPath = "";
+
+  // Only add the root path if one does not already exist for Location.
+  if( Location.find( ':' ) == StringRef::npos )
+    fullPath = m_rootPath;
+
+  fullPath += Location + "(" + std::to_string(lineNumber) + ")"; 
+
+  std::replace(fullPath.begin(), fullPath.end(), '\\', '/');
+
+  return fullPath;
+}
+
+void LiveValueAnalysis::analyzeCFG() {
+  
+  // Find TraceRay call sites.
+  for( Function &main : *m_module ) {
+    for( Function &F : *m_module ) {
+
+      if( !F.getName().startswith( "dx.op.traceRay" ) )
+        continue;
+
+      for( User *user : F.users() ) {
+        CallInst *call = dyn_cast<CallInst>(user);
+        if( call && call->getParent()->getParent() == &main )
+          m_callSites.insert( call );
+      }
+    }
+  }
+
+  // Determine live values across each call.
+  std::set<BasicBlock *> phiSelectorSpills;
+
+  // Prepare Debug Info.
+  DebugInfoFinder DIF;
+  DIF.processModule(*m_module);
+
+  for (CallInst *callSite : m_callSites) {
+    // Determine CFG edges that can lead to this call site.
+    std::set<std::pair<Instruction *, Instruction *>> predCfgEdges;
+    {
+      BasicBlock *BB = callSite->getParent();
+      std::vector<BasicBlock *> worklist;
+
+      for (;;) {
+        for (auto i = llvm::pred_begin(BB); i != pred_end(BB); ++i) {
+          BasicBlock *pred = *i;
+          std::pair<Instruction *, Instruction *> edge =
+            std::make_pair(pred->getTerminator(), BB->begin());
+
+          if (!predCfgEdges.count(edge)) {
+            predCfgEdges.insert(edge);
+            worklist.push_back(pred);
+          }
+        }
+
+        if (worklist.empty())
+          break;
+
+        BB = worklist.back();
+        worklist.pop_back();
+      }
+    }
+
+    // Determine known branch conditions based on CFG edges.
+    // Prunes pred CFG edges based on invariants iteratively.
```

Comment:
IIUC then this is what happens during this "pruning":
- We have collected all edges which can be taken from the start of the function to the current call site (predCfgEdges).
- Look at the branch of each of these edges in random order and check if either of the outgoing edges can be removed. This is super confusing already: We can actually end up removing a different edge from the one we're looking at. This is also the reason for the "cast" mentioned by Greg below. It's not actually a cast, it's a copy of the entire set. And it's required because otherwise we could be invalidating the iterator when erasing a value from the set we're iterating over. That's also the reason for the "if (!count) continue" case, which intuitively you'd expect to be impossible to hit, but it's possible because we are iterating over a set that's different from the one we're testing against.
- If an edge was removed, start iterating over the remaining set of edges in random order again. We should really be using a proper CFG traversal, this should be possible in a single pass.
- We only ever remove an edge if it has a predecessor that has multiple outgoing edges (exactly 2, actually). This means we will at most remove individual edges and never entire paths. The information in predCfgEdges is later used to determine if an incoming value to a phi has to be taken into account (if the edge is not in predCfgEdges it can be ignored). This means that the only case when removal of an edge has an effect is when the target included a phi, which means the edge was critial (if we ran BreakCriticalEdges before, which I hope, then we'll never have any effect of this).

Actually, I'm starting to think you don't need any of this and could just use a set of the blocks which can be traversed on the path from start to call site. If later you want to know which incoming values of a phi have to be taken into account, just check if the corresponding incoming blocks are part of that set.

I'm probably missing some things, though... for example I'm now also starting to question under which circumstances you could ever hit the count == 2 case AND have the condition marked as "always true" or "always false" already... maybe only if it's a condition shared by multiple conditionals?

(parts of this, if accurate, is really what should be in the comment btw).